### ISSUE TITLE
consume monorepo packages in consumer without publishing a new version

### ISSUE THREAD

**Issue by gaurav5430** [lerna/lerna#2519]
## Expected Behavior
I am using lerna to maintain a UI component library, which is consumed by other projects. Whenever i make some changes to a component, I test it locally by symlinking my monorepo in one of the consumer project, this works fine for local testing. 
Now, when i need to host this code somewhere for the QA to test, I need to be able to refer to the latest changes in the component monorepo. The way I do that currently is, I publish a version, consume that version in the consumer project and then host the consumer project on a url for the QA to test. The problem with this approach is that the changes to the component are not yet finalised before QA approval, but in the meantime I would have published a few versions which are broken and I don't want anyone to consume.

## Possible Solution
I tried to npm install from a git repo instead of a published version, but there is currently no possibility to refer a subdirectory inside git in npm install(https://github.com/npm/npm/issues/2974)

All in all, I am looking for a way to consume lerna monorepo packages without publishing a version, or to figure out a way to test a package without publishing multiple times.

**Comment by JabbyPanda** [comment]
We consume shared library w/o publishing it to NPM in Lerna monorepo by mentioning it in "dependencies" section of 'package..json' of client projects

We run "lerna bootstrap" to symlink this shared library when we publish any client project to NPM

**Comment by gaurav5430** [comment]
Apologies, I don't get it. 
Is your consumer/client part of the monorepo?

**Comment by JabbyPanda** [comment]
> Is your consumer/client part of the monorepo?

Yes

**Comment by gaurav5430** [comment]
Yeah, in the workflow that I am talking about, the monorepo is the ui component library, and the consumer is not part of this monorepo. 
The conusmer needs to have access to latest changes on the monorepo without publishing, so that the monorepo changes can be tested before they get published

**Comment by JabbyPanda** [comment]
we consume shared library code as npm dependency in monorepo like this

```
{
  "name": "our-view",
  "dependencies": {
     "shared-dependency": "^1.0.0"
}
```

shared-dependency **is never published to npm** - it is symlinked by running  _lerna bootstrap_ command

This setup is not ideal, but works pretty good.

**Comment by gaurav5430** [comment]
Sure, my use case seems completely different from yours then.

My consumer is not part of the monorepo at all. There is nothing that lerna bootstrap can help with in this setup. Probably it would make sense for me to publish/unpublish alpha/beta versions for internal testing

**Comment by powercoder23** [comment]
Have you tried 
[yarn link](https://classic.yarnpkg.com/en/docs/cli/link/)
or 
[npm link](https://docs.npmjs.com/cli/link.html)

**Comment by gaurav5430** [comment]
@powercoder23 
As mentioned in my original description, i do use symlinking when testing on local, but that doesn't work when I need to deploy the consumer project to a server so that others (QA) can play with it. 
I mean, it would be quite difficult to make symlinking setup work on the server. first we would need to make sure that this particular branch of monorepo available on the server, then use a script to symlink it to the consumer, then do npm install ...

**Comment by karangarg45** [comment]
you can use the file syntax in package.json. Something like this
```
{
  "name": "baz",
  "dependencies": {
    "bar": "file:../foo/bar"
  }
}
```
On your jenkins or bamboo server or whatever CI/CD tool you use for deployments, you will need to clone both your monorepo and the app code. Then you can have a shell script which does all the replacement in your package.json with the relative paths of monorepo packages inside the npm repo. Then npm install will do the rest of job.

**Comment by gaurav5430** [comment]
> you can use the file syntax in package.json. Something like this
> 
> ```
> {
>   "name": "baz",
>   "dependencies": {
>     "bar": "file:../foo/bar"
>   }
> }
> ```
> 
> On your jenkins or bamboo server or whatever CI/CD tool you use for deployments, you will need to clone both your monorepo and the app code. Then you can have a shell script which does all the replacement in your package.json with the relative paths of monorepo packages inside the npm repo. Then npm install will do the rest of job.

Yup, that is what I am doing locally, and this is what i am trying to avoid on CI as this is too much work. I am not opposed to this though if there is no better solution.

**Comment by karangarg45** [comment]
can't you make your app as the part of your monorepo?

**Comment by gaurav5430** [comment]
component library monorepo developers and consumers of the component library are different teams altogether, so this is not feasible.

**Comment by karangarg45** [comment]
ohh i get it now..i haven't explored Module federation much but seems like a good candidate for your problem. May be you can explore more on that front.

**Comment by KKS1** [comment]
Curious to know if anyone has found a solution to this extremely common but combersome problem? If for app and ui library sitting inside same monorepo, this mystery remains, where while deploying for QA testing, we have to put up a new version say rc- releases. 
IMP Note: The file path local dependencies break the symlinking when using usual lerna plus year workspaces locally. Thus, using it for local development actually makes process more non-traditional. (Correct me if someone has a work around here)

**Comment by gaurav5430** [comment]
we have resorted to publishing beta versions for testing and then unpublishing them :(, as we publish everything on our internal npm registry and we know who are the consumers and that they wouldn't be using the beta versions

not ideal, but works for our small team setup

**Comment by KKS1** [comment]
Thanks for putting in your inputs @gaurav5430. Yeah that's what I meant by rc- releases, and then resorting to delete them later once QA done. A lot of manual process but I guess most of us will have to follow this only, till community finds a better solution.

**Comment by quannt** [comment]
I have the exact use case and yeah, nothing works out of the box. I guess the behavior makes sense from the consumer (the application) point of view, but it would be nice if there's a way to make this work.

**Comment by KKS1** [comment]
Totally agree!

**Comment by oliviera9** [comment]
> I have the exact use case and yeah, nothing works out of the box. I guess the behavior makes sense from the consumer (the application) point of view, but it would be nice if there's a way to make this work.

I'm in the same boat. No solution yet. The only thing I noticed is that when installing the monorepo from the GitHub repository, the `prepare` script gets called, so I was trying to figure out if something could be done from that script.

**Comment by JamesHenry** [comment]
Hi Folks, I'm going to close this older issue. We have lots of different options available to use these days like https://www.verdaccio.org/ (local or self-hosted) https://pkg.pr.new/ etc and this is not in scope for Lerna directly.
